app-frontend/src/app/components/diagramContainer/diagramContainer.controller.js

```diff
@@ -229,6 +132,205 @@ export default class DiagramContainerController {
         });
     }
 
+    extractInputs() {
+        this.inputsJson = [];
+
+        let json = Object.assign({}, this.toolJson);
+        let inputs = [json];
+        while (inputs.length) {
+            let input = inputs.pop();
+            let args = input.args;
+            if (args) {
+                let tool = this.getToolLabel(input);
+                if (!Array.isArray(args)) {
+                    args = Object.values(args);
+                }
+                inputs = inputs.concat(args.map((a) => {
+                    return Object.assign({parent: tool}, a);
+                }));
+            } else {
+                this.inputsJson.push(input);
+            }
+        }
+    }
+
+    extractShapes() {
+
+        function shouldCreateRect(input) {
+            // We only create nodes for non-inputs or layer inputs
+            let args = input.args;
+            return args || !args && input.type === 'layer';
+        }
+
+        let nextId = 0;
+        let nodes = new Map();
+        let shapes = [];
+        let json = Object.assign({}, this.toolJson);
+        let inputs = [json];
+
+        while (inputs.length) {
+            let input = inputs.pop();
+            let rectangle;
+            let args;
+
+            // Args can be array or object, if object, convert to array
+            if (input.args) {
+                args = Array.isArray(input.args) ? input.args : Object.values(input.args);
+            } else {
+                args = [];
+            }
+
+
+            if (shouldCreateRect(input)) {
```

Comment:
We should just inline the comparison - I broke it out because I thought I'd be using it more than once